### ISSUE TITLE
CIFAR-10 img classification

### ISSUE THREAD

**Issue by songhee-cho** [songhee-cho/ML-study-biplug#1]
### **학습 목적**🍐

* **CIFAR-10 data set을 10개의 label에 따라 classification**

### **model의 layer 구성**🧙🏻

model의 layer를 도식화한 표를 첨부해주세요.

### **model을 구현하기 위한 코드**🧙🏻
* **_Pytorch 사용 학습 환경 구성_**

**1) Import Libs**

```python
#Google Colab에서 Pytorch 사용

import torch
import matplotlib.pyplot as plt
import numpy as np

from torch import nn
from torch.nn import functional as F
from torchvision import datasets, transforms
```

**2) 구글 드라이브 환경에서 사용할 수 있도록 지정**

```python
from google.colab import drive
drive.mount('/content/gdrive')
```

**3) device 확인**

```python
#사용 device 확인: GPU(cuda) or CPU
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
device
```

**4) CIFAR 10 data set 다운로드(training set, test set 구분), augmentation, normalization**

CIFAR 10을 활용한 모델 정확도를 높이기 위해, 다양한 data augmentation을 실시했습니다.

- RandomHorizontalFlip(): 이미지를 수평 방향으로 뒤집기
- RandomRoation(degrees): 이미지를 degrees 각도 이하로 회전
- RandomAffine(): Affine transformation

아핀 변환은 **점, 직선, 평면을 보존하는 선형 매핑 방법**입니다. 아핀 변환 후에도 평행한 선들은 평행한 상태로 유지됩니다.

아핀 변환 기술은 **이상적이지 않은 카메라 각도에서 발생하는 기하학적 왜곡이나 형태 변형을 보정하는 데 주로 사용**됩니다. 영상 왜곡을 고려하지 않고 상호 작용 및 계산을 쉽게 수행할 수 있습니다.

```python
# dataset upload
# data augmentation
# data normalization

transform_train = transforms.Compose([transforms.Resize((32, 32)),
                                      transforms.RandomHorizontalFlip(),
                                      transforms.RandomRotation(10),
                                      transforms.RandomAffine(0, shear=10, scale=(0.8, 1.2)),
                                      transforms.ColorJitter(brightness=0.2, contrast=0.2, saturation=0.2),
                                      transforms.ToTensor(),
                                      transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
                                     ])

transformer = transforms.Compose([transforms.Resize((32, 32)),
                                  transforms.ToTensor(),
                                  transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
                                 ])

training_dataset = datasets.CIFAR10(root='./data', train=True, download=True, transform=transform_train)
test_dataset = datasets.CIFAR10(root='./data', train=False, download=True, transform=transformer)

training_loader = torch.utils.data.DataLoader(dataset=training_dataset, batch_size=100, shuffle=True)
test_loader = torch.utils.data.DataLoader(dataset=validation_dataset, batch_size=100, shuffle=False)
```

**5) 임의의 data set 출력**

```python
def im_convert(tensor):
  image = tensor.clone().detach().numpy()
  image = image.transpose(1, 2, 0)
  image = image * np.array([0.5, 0.5, 0.5] + np.array([0.5, 0.5, 0.5]))
  image = image.clip(0, 1)
  return image

classes = ("plane", "car", "bird", "cat", "deer", "dog", "frog", "horse", "ship", "truck")

dataiter = iter(training_loader)
images, labels = dataiter.next()

fig = plt.figure(figsize=(25, 4))

for i in np.arange(20):
  # row 2 column 10
  ax = fig.add_subplot(2, 10, i+1, xticks=[], yticks=[])
  plt.imshow(im_convert(images[i]))
  ax.set_title(classes[labels[i].item()])
```

**6) hyperparameter 정의**

```python
# hyperparmeter

num_epoch = 15
learning_rate = 0.001
```

**7) CNN을 활용한 classifier model 정의**

MNIST는 숫자 이미지 data set이지만, CIFAR 10은 사물 이미지이므로 accuracy를 확보하기가 어렵습니다.

따라서, data augmentation과 함께, model 구성 시 3개의 conv layer 이외에도 **drop out** 기법을 추가적으로 사용했습니다. 

```python
# model
class LeNet_modified(nn.Module):

  def __init__(self):
    super().__init__()
    # 32px 이었다가, conv를 거치면서 절반으로 감소
    self.conv1 = nn.Conv2d(3, 16, 3, 1, padding=1)
    self.conv2 = nn.Conv2d(16, 32, 3, 1, padding=1)
    self.conv3 = nn.Conv2d(32, 64, 3, 1, padding=1)
    self.fc1 = nn.Linear(4*4*64, 500)
    self.dropout1 = nn.Dropout(0.5)
    self.fc2 = nn.Linear(500, 10)

  def forward(self, x):
    x = F.relu(self.conv1(x))
    x = F.max_pool2d(x, 2, 2)
    x = F.relu(self.conv2(x))
    x = F.max_pool2d(x, 2, 2)
    x = F.relu(self.conv3(x))
    x = F.max_pool2d(x, 2, 2)
    x = x.view(-1, 4*4*64)
    x = F.relu(self.fc1(x))
    x = self.dropout1(x)
    x = self.fc2(x)
    return x

  
model = LeNet_modified()
```

**8) 학습 시 optimizer와 loss를 판단하는 criterion 정의: 다중 class에 적합한 cross-entropy, 일반적으로 사용하는 Adam optimizer 활용**

```python
# optimizer and criterion
import torch.optim as optim

criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
```

```python
#training
epochs = num_epoch
running_loss_history = []
running_correct_history = []
test_running_loss_history = []
test_running_correct_history = []

for e in range(epochs):

  running_loss = 0.0
  running_correct = 0.0
  test_running_loss = 0.0
  test_running_correct = 0.0

  for inputs, labels in training_loader:

    inputs = inputs.to(device)
    labels = labels.to(device)
    outputs = model(inputs)
    loss = criterion(outputs, labels)

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    _, preds = torch.max(outputs, 1)

    running_correct += torch.sum(preds == labels.data)
    running_loss += loss.item()

  else:
    # 훈련팔 필요가 없으므로 메모리 절약
    with torch.no_grad():

      for test_input, test_label in test_loader:

        test_input = test_input.to(device)
        test_label = test_label.to(device)
        test_outputs = model(test_input)
        test_loss = criterion(test_outputs, test_label)

        _, test_preds = torch.max(test_outputs, 1)
        test_running_loss += test_loss.item()
        test_running_correct += torch.sum(test_preds == test_label.data)

    epoch_loss = running_loss / len(training_loader)
    epoch_acc = running_correct.float() / len(training_loader)
    running_loss_history.append(epoch_loss)
    running_correct_history.append(epoch_acc)

    test_epoch_loss = test_running_loss / len(test_loader)
    test_epoch_acc = test_running_correct.float() / len(test_loader)
    test_running_loss_history.append(test_epoch_loss)
    test_running_correct_history.append(test_epoch_acc)

    print("===================================================")
    print("epoch: ", e + 1)
    print("training loss: {:.5f}, acc: {:5f}".format(epoch_loss, epoch_acc))
```

**10) Google drive에 .pt 저장**

선행되어야 할 사항: 먼저 자신의 drive에 'model' 이름으로 폴더 생성

```python
#Google drive에 폴더명: model인 폴더 생성
# 경로 지정
PATH =  "/content/gdrive/My Drive/model/model_cifar10.pt"
torch.save(model.state_dict(), PATH)
```

**11) test data로 성능 확인**

```python
# test data로 확인
model.load_state_dict(torch.load(PATH))

# 전체 test dataset으로 모델 정확도 확인

correct = 0
total = 0
with torch.no_grad():
    for data in test_loader:
        images, labels = data
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()

print('Accuracy of the network on the 10000 test images: %d %%' % (
    100 * correct / total))
```

**12) 카테고리에 따른 정확도 확인**

```python
#label별 정확도 확인

class_correct = list(0. for i in range(10))
class_total = list(0. for i in range(10))
with torch.no_grad():
    for data in test_loader:
        images, labels = data
        outputs = model(images)
        _, predicted = torch.max(outputs, 1)
        c = (predicted == labels).squeeze()
        for i in range(4):
            label = labels[i]
            class_correct[label] += c[i].item()
            class_total[label] += 1

for i in range(10):
    print('Accuracy of %5s : %2d %%' % (
        classes[i], 100 * class_correct[i] / class_total[i]))
```

```python
#플롯설정 
fig = plt.figure(figsize=(18,14))
columns = 5
rows = 5

model.eval()
cnt = 0 
total_cnt =columns*rows
for i in range(1, columns*rows+1):
    #랜덤으로 테스트데이터에서 이미지 픽업 
    data_idx = np.random.randint(len(test_dataset))
    input_img = test_dataset[data_idx][0].unsqueeze(dim=0).to(device) 
 
    output = model(input_img)    
    _, argmax = torch.max(output, 1)
    
    
    #print(output)
    #output 에서 몇번째 값이 제일 확률적으로 높은지 표시    
    #print(argmax.item())
    
    pred = classes[argmax.item()]
    label = classes[test_dataset[data_idx][1]]
    
    fig.add_subplot(rows, columns, i)
    if pred == label:
        plt.title(pred + ', right !!')
        cmap = 'Blues'
        cnt = cnt +1
        #print("right")
    else:
        plt.title('Not ' + pred + ' but ' +  label)
        cmap = 'Reds'
        #print("error")
    plot_img = test_dataset[data_idx][0][0,:,:]
    plt.imshow(plot_img, cmap=cmap)
    plt.axis('off')
    
plt.show() 

print('Accuracy for {} images: {:.2f}%'.format(total_cnt, cnt / total_cnt * 100))
```

```python
torch.save({
            'epoch': num_epoch,
            'model_state_dict': model.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
            'loss': loss
            }, "/content/gdrive/My Drive/model/model_cifar10.pt")

torch.save(model.state_dict(), "/content/gdrive/My Drive/model/model_cifar10_min.pt")

checkpoint = torch.load(PATH)

model.load_state_dict(checkpoint['model_state_dict'])
optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
epoch = checkpoint['epoch']
loss = checkpoint['loss']

model.eval()
```

### **학습 그래프**🧠

구현한 모델을 통해 학습한 결과 그래프를 accuracy 등과 함께 첨부해주세요.

### 문제점 **🤔**

모델 학습 시 문제점 및 개선 사항이 있다면 적어주세요.
- **참고한 코드 출처**

[Pytorch 08) - CIFAR 10 학습](https://yceffort.kr/2019/02/25/pytorch-08-CIFAR-10)
